4x4 Board

damonkelley opened the pull request:
I started with some refactoring to make it easier to add the new board size. I merged the `Game` and `GameRules`, move the `nextMove()` behavior to the `GameLoop`, and removed the circular dependency between `Game` and `Players`.

This ended up being a more involved change than I had anticipated (or estimated for).

**Caveat**:
The minimax algorithm will need further tuning to always win a  4x4 game. Based on my testing, `minimax` with a depth of 4 will win 85% of paths through the game. Using the current implementation, and depth greater than 4 will take much too long, and the effect on the usability is too great. A depth of 3 will take the computer ~3 seconds on the first or second move, while a depth of 5 will take ~30 or more. I have started looking to ways of optimizing the algorithm beyond tuning the depth, like alpha-beta pruning.

damonkelley committed:
Merge the Game and GameRules classes

damonkelley committed:
Make the State object unknown to clients of the Game class

damonkelley committed:
Remove the toString debug helper code

damonkelley committed:
Remove getWinner and make canMove private

damonkelley committed:
Import the FakeUI into the GameLoopTest

damonkelley committed:
Remove explicit calls to the nextMove method

damonkelley committed:
Introduce the Players class

damonkelley committed:
Update the GameLoop let players move on the game

damonkelley committed:
Create a single NullFinder class

damonkelley committed:
Update the Game so that it no longer requires any Players

damonkelley committed:
Remove the GameBuilder

Now that the Game only requires a single parameter, the GameBuilder is
no longer necessary.

damonkelley committed:
Remove iml files

damonkelley committed:
Add the BoardFormatter

The BoardFormatter is responsible for formatting the board into rows,
columns, and diagonals.

damonkelley committed:
Update the board to accept a size

The size will determine the width and height of the board.
Creating a board with no size will result in a 3x3 board.

damonkelley committed:
Allow the game to be configured with a 3x3 or a 4x4 board from the UI

damonkelley committed:
Modify the GamePresenter to handle a 4x4 board

damonkelley committed:
Optimize the ArtificialIntelligenceFinder for a 4x4 board

damonkelley committed:
Update the config menu to ask for the board size first

damonkelley committed:
Allow the HumanFinder to accept input for a 4x4 board

damonkelley committed:
Update the BoardFormatter to collect clusters

damonkelley committed:
Fix a regression in the Board

After the size was introduced, boards that were copied were assigned the
default size instead of the size of the original board.

damonkelley committed:
Fix a Game regression

A Game that was reset was being created with the default board size of
three instead of using the size of the original.

Nayshins commented on src/main/java/me/damonkelley/tictactoe/BoardFormatter.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    public List<List<Marker>> collect() {
+        List<List<Marker>> collection = new ArrayList<>();
```
I think using the name collection here may be a bit confusing because collection is the interface which list is implementing.

Nayshins commented on src/main/java/me/damonkelley/tictactoe/BoardFormatter.java:
```diff
@@ -0,0 +1,109 @@
+        return groupSpacesBy(this::getX);
+    }
+
+    private List<List<Marker>> groupSpacesBy(Function<Map.Entry<Space, Marker>, Integer> getY) {
```
I'm having a hard time following what is going on in this method. It looks to me like there are 2-3 distinct stream operations that are going on here.

damonkelley replied:
I agree. I am beginning to wonder if calculating the the possible wins is really worth it right now. Perhaps I introduced this too early.

Nayshins commented on src/main/java/me/damonkelley/tictactoe/Players.java:
```diff
@@ -0,0 +1,19 @@
+        players = Arrays.asList(playerA, playerB);
+    }
+
+    public Player get(Marker x) {
```
I think using `x` as your variable name here is unclear. Is the stream always filtering for the player with the `X` marker? I would prefer a more clear variable name here.

Nayshins commented on src/main/java/me/damonkelley/tictactoe/Players.java:
```diff
@@ -0,0 +1,19 @@
+        return players.stream()
+                .filter(player -> player.getMarker() == x)
+                .findFirst()
+                .get();
```
Be careful of using get directly on a stream because it can throw a `NoSuchElementException` if nothing exists.

Nayshins commented on src/main/java/me/damonkelley/tictactoe/Game.java:
```diff
@@ -2,65 +2,83 @@
-            return player1;
-        } else if (winningMarker == player2.getMarker()) {
-            return player2;
+    public Marker determineWinner() {
```
I feel like there is an opportunity to trim this conditional statement. Maybe there is another response that could be returned instead of null that would prevent you from having to do null checks in other methods?

damonkelley commented on src/main/java/me/damonkelley/tictactoe/BoardFormatter.java:
```diff
@@ -0,0 +1,109 @@
+                .filter(e -> e.getKey().y < 3 && e.getKey().x < 3)
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+    }
```
I found a bug somewhere in here I believe

damonkelley committed:
Update the formatter so that it gets the diagonals based on board size

damonkelley committed:
Optimize the minimax algorithm using alpha beta pruning

With this optimization, the depth can be increased to 7 for 3x3 and 4x4
boards while keeping the game usable from the user's perspective.

trptcolin commented on src/main/java/me/damonkelley/ui/GamePresenter.java:
```diff
@@ -2,63 +2,90 @@
-        for (Space space : game.getBoard()) {
-            addSpace(space);
-            addSeparator();
+        String template = (game.getBoard().getSize() == 4)? FOUR_BY_FOUR : THREE_BY_THREE;
```
It's a little funky that you're making this decision every time you present the board - seems like something that you could just do once, perhaps in the constructor?

trptcolin commented on src/main/java/me/damonkelley/ui/GamePresenter.java:
```diff
@@ -2,63 +2,90 @@
+                                          "\n";
+
     public GamePresenter(Game game) {
         this.game = game;
```
As far as I can tell, this class only uses the `Game` to get ahold of a `Board`. So you end up with some unneeded dependencies - a little extra knowledge that you could eliminate by only taking in what you need. See also [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter).

trptcolin commented on src/main/java/me/damonkelley/tictactoe/finder/ArtificialIntelligenceFinder.java:
```diff
@@ -1,61 +1,74 @@
+        return depth;
     }
 
     private Integer findMinimumScore(HashMap<Integer, Space> scores) {
```
It looks like this data structure isn't doing you any favors in terms of read performance. You might consider looking at a sorted tree data structure like [`TreeMap`](https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html) so that you're more like O(1) than O(N) in finding the max/min score.

trptcolin commented on src/test/java/me/damonkelley/tictactoe/PlayersTest.java:
```diff
@@ -0,0 +1,17 @@
+        Player playerB = new Player(Marker.O, null);
+
+        Players players = new Players(playerA, playerB);
+        assertTrue("It can get Player A", playerA == players.get(Marker.X));
```
1. Be careful with `==` on instances of `java.lang.Object` - prefer `.equals` pretty much all the time - reference equality is a dangerous habit to get into. On the other hand, it's fine here.
2. `assertTrue` generally gives garbage error messages like "Expected true, got false". Prefer `assertEquals`, or if you reaalllly want reference equality, maybe `assertSame`?

damonkelley committed:
Update the minimax algorithm to use a TreeMap over a HashMap

damonkelley committed:
Update the PlayerTest assertions to use assertSame

damonkelley committed:
Update the signature of the get method to be more informative

damonkelley committed:
Convert the GamePresenter into the BoardPresenter

damonkelley committed:
Replace determineWinner with isWinner and hasWinner

damonkelley committed:
Utilize the TreeMap API in finding the min and max scores

damonkelley committed:
Increase the depth of minimax to 8

damonkelley committed:
Convert the BoardFormatter into the Partitioner

damonkelley committed:
Move the 4x4 functional tests to a separate test case

damonkelley committed:
Add Travis configuration

damonkelley commented:
As of 29bf5ef, the `ArtificialIntelligenceFinder` should win the game 100% of the time.

trptcolin commented:
👍  LGTM
